identity/extension_verification.go

```diff
@@ -24,40 +32,58 @@ func NewSchemaExtensionVerification(i *Identity, lifespan time.Duration) *Schema
 	return &SchemaExtensionVerification{i: i, lifespan: lifespan}
 }
 
+const (
+	ChannelTypeEmail = "email"
+	ChannelTypeSMS   = "sms"
+	ChannelTypePhone = "phone"
+)
+
 func (r *SchemaExtensionVerification) Run(ctx jsonschema.ValidationContext, s schema.ExtensionConfig, value interface{}) error {
 	r.l.Lock()
 	defer r.l.Unlock()
 
-	switch s.Verification.Via {
-	case AddressTypeEmail:
-		if !jsonschema.Formats["email"](value) {
-			return ctx.Error("format", "%q is not valid %q", value, "email")
-		}
-
-		address := NewVerifiableEmailAddress(
-			strings.ToLower(strings.TrimSpace(
-				fmt.Sprintf("%s", value))), r.i.ID)
-
-		r.appendAddress(address)
+	if s.Verification.Via == "" {
+		return nil
+	}
 
+	format, ok := s.RawSchema["format"]
+	if !ok {
+		format = ""
+	}
+	formatString, ok := format.(string)
+	if !ok {
 		return nil
+	}
 
-	case AddressTypePhone:
-		if !jsonschema.Formats["tel"](value) {
-			return ctx.Error("format", "%q is not valid %q", value, "phone")
+	if formatString == "" {
+		switch s.Verification.Via {
+		case ChannelTypeEmail:
+			formatString = "email"
+			formatter, ok := jsonschema.Formats[formatString]
+			if !ok {
+				supportedKeys := maps.Keys(jsonschema.Formats)
+				return ctx.Error("format", "format %q is not supported. Supported formats are [%s]", formatString, strings.Join(supportedKeys, ", "))
+			}
+
+			if !formatter(value) {
+				return ctx.Error("format", "%q is not valid %q", value, formatString)
+			}
```

Comment:
This block needs to be moved to after the if, right?

Reply:
Ah I guess not, because the library already did the check and we don't want to double-validate.

Reply:
yup, exactly, the library doesn't de-duplicate the error messages, so this would lead to ugly errors.